src/main/java/org/springframework/data/domain/Range.java

```diff
@@ -68,13 +63,127 @@ public Range(T lowerBound, T upperBound) {
 	 * @param upperBound can be {@literal null}.
 	 * @param lowerInclusive
 	 * @param upperInclusive
+	 * @deprecated since 2.0. Use {@link Range#of(Bound, Bound)} and {@link Bound} factory methods:
+	 *             {@link Bound#inclusive(Comparable)}, {@link Bound#exclusive(Comparable)}/{@link Bound#unbounded()}.
 	 */
+	@Deprecated
 	public Range(T lowerBound, T upperBound, boolean lowerInclusive, boolean upperInclusive) {
 
+		this.lowerBound = lowerBound == null ? Bound.unbounded()
+				: lowerInclusive ? Bound.inclusive(lowerBound) : Bound.exclusive(lowerBound);
+
+		this.upperBound = upperBound == null ? Bound.unbounded()
+				: upperInclusive ? Bound.inclusive(upperBound) : Bound.exclusive(upperBound);
+	}
+
+	private Range(Bound<T> lowerBound, Bound<T> upperBound) {
+
+		Assert.notNull(lowerBound, "Lower boundary must not be null!");
+		Assert.notNull(upperBound, "Upper boundary must not be null!");
+
 		this.lowerBound = lowerBound;
 		this.upperBound = upperBound;
-		this.lowerInclusive = lowerInclusive;
-		this.upperInclusive = upperInclusive;
+	}
+
+	/**
+	 * Returns an unbounded {@link Range}.
+	 * 
+	 * @return
+	 * @since 2.0
+	 */
+	@SuppressWarnings("unchecked")
+	public static <T extends Comparable<T>> Range<T> unbounded() {
+		return (Range<T>) UNBOUNDED;
+	}
+
+	/**
+	 * Create a {@link RangeBuilder} given the lower {@link Bound}.
+	 * 
+	 * @param lower must not be {@literal null}.
+	 * @return
+	 * @since 2.0
+	 */
+	public static <T extends Comparable<T>> RangeBuilder<T> from(Bound<T> lower) {
+
+		Assert.notNull(lower, "Lower bound must not be null!");
+		return new RangeBuilder<>(lower);
+	}
+
+	/**
+	 * Creates a new {@link Range} with the given lower and upper bound.
+	 * 
+	 * @param lowerBound must not be {@literal null}.
+	 * @param upperBound must not be {@literal null}.
+	 * @since 2.0
+	 */
+	public static <T extends Comparable<T>> Range<T> of(Bound<T> lowerBound, Bound<T> upperBound) {
+		return new Range<>(lowerBound, upperBound);
+	}
+
+	/**
+	 * Creates a new {@link Integer} {@link Range} with the given lower and upper bound. Treats the given values as
+	 * including bounds. Use {@link #of(Bound, Bound)} to configure different bound behavior.
+	 * 
+	 * @param lowerBound
+	 * @param upperBound
+	 * @since 2.0
+	 */
+	public static Range<Integer> from(int lowerBoundInclusive, int upperBoundInclusive) {
```

Comment:
Can we remove those `from(…)` methods taking two arguments as they're effectively aliasing the factory methods on `Bound`, don't they?